File: breadth-first-search/count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0 
         
        q = collections.deque()
        self.count_good = 0
        # (cur_node, max value from root to cur_node)
        q.append((root,-pow(10,4))) 
        while q: 
            cur,tmp_max = q.popleft()
            if cur.val >= tmp_max: 
                tmp_max = cur.val
                self.count_good+=1
            if cur.left is not None:
                q.append((cur.left,tmp_max)) 
            if cur.right is not None: 
                q.append((cur.right,tmp_max)) 
        return self.count_good



            
                    
        

File: depth-first-search/count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int count = 1; 

    public int goodNodes(TreeNode root) {
        if (root == null) return 0; 
        dfs(root, root.val); 
        return this.count; 
    }

    void dfs(TreeNode root, int max){
        if (root.left != null){
            if (root.left.val >= max){
                this.count++; 
                dfs(root.left, root.left.val);
            }else{
                dfs(root.left, max); 
            }
        }
        if (root.right != null){
            if (root.right.val >= max){
                this.count++; 
                dfs(root.right, root.right.val); 
            }else{
                dfs(root.right, max); 
            }
        }
    }
}